File: dbm-ui/backend/db_periodic_task/local_tasks/db_meta/db_meta_check/mysql_cluster_topo/tendbha/status.py
# -*- coding: utf-8 -*-
"""
TencentBlueKing is pleased to support the open source community by making 蓝鲸智云-DB管理系统(BlueKing-BK-DBM) available.
Copyright (C) 2017-2023 THL A29 Limited, a Tencent company. All rights reserved.
Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
You may obtain a copy of the License at https://opensource.org/licenses/MIT
Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
"""

from typing import List

from django.utils.translation import gettext_lazy as _

from backend.db_meta.enums import ClusterEntryRole, ClusterStatus, InstanceInnerRole, InstancePhase, InstanceStatus
from backend.db_meta.models import Cluster
from backend.db_periodic_task.local_tasks.db_meta.db_meta_check.mysql_cluster_topo.check_response import CheckResponse
from backend.db_periodic_task.local_tasks.db_meta.db_meta_check.mysql_cluster_topo.decorator import checker_wrapper
from backend.db_report.enums import MetaCheckSubType


@checker_wrapper
def _cluster_status(c: Cluster) -> List[CheckResponse]:
    if c.status != ClusterStatus.NORMAL:
        return [
            CheckResponse(
                msg=_("集群状态异常: {}".format(c.status)),
                check_subtype=MetaCheckSubType.ClusterTopo,
            )
        ]


@checker_wrapper
def _cluster_instance_status(c: Cluster) -> List[CheckResponse]:
    bad = []
    for si in c.storageinstance_set.all():
        if si.status != InstanceStatus.RUNNING or si.phase != InstancePhase.ONLINE:
            bad.append(
                CheckResponse(
                    msg=_("实例 {} 状态异常: {}, {}".format(si.ip_port, si.status, si.phase)),
                    check_subtype=MetaCheckSubType.ClusterTopo,
                    instance=si,
                )
            )

    for pi in c.proxyinstance_set.all():
        if pi.status != InstanceStatus.RUNNING or pi.phase != InstancePhase.ONLINE:
            bad.append(
                CheckResponse(
                    msg=_("实例 {} 状态异常: {}, {}".format(pi.ip_port, pi.status, pi.phase)),
                    check_subtype=MetaCheckSubType.ClusterTopo,
                    instance=pi,
                )
            )

    return bad


@checker_wrapper
def _cluster_master_entry_count(c: Cluster) -> List[CheckResponse]:
    """
    至少 1 个主访问入口
    """
    cnt = 0
    for ce in c.clusterentry_set.all():
        if ce.role == ClusterEntryRole.MASTER_ENTRY:
            cnt += 1

    if cnt <= 0:
        return [CheckResponse(msg=_("缺少主访问入口"), check_subtype=MetaCheckSubType.ClusterTopo)]


@checker_wrapper
def _cluster_proxy_count(c: Cluster) -> List[CheckResponse]:
    """
    至少 2 个存活的 proxy
    """
    cnt = 0
    for pi in c.proxyinstance_set.all():
        if pi.status == InstanceStatus.RUNNING and pi.phase == InstancePhase.ONLINE:
            cnt += 1

    if cnt < 2:
        return [CheckResponse(msg=_("正常 proxy 不足 2 个"), check_subtype=MetaCheckSubType.ClusterTopo)]


@checker_wrapper
def _cluster_one_master(c: Cluster) -> List[CheckResponse]:
    """只能有一个 master"""
    m = []
    for si in c.storageinstance_set.all():
        if si.instance_inner_role == InstanceInnerRole.MASTER:
            m.append(si)

    if len(m) <= 0:
        return [CheckResponse(msg=_("无 master 实例"), check_subtype=MetaCheckSubType.ClusterTopo)]

    if len(m) > 1:
        return [
            CheckResponse(
                msg=_("master 多余 1 个: {}".format(",".join([ele.ip_port for ele in m]))),
                check_subtype=MetaCheckSubType.ClusterTopo,
            )
        ]


@checker_wrapper
def _cluster_master_status(
    c: Cluster,
) -> List[CheckResponse]:
    """
    master 必须
    status == running
    phase == online
    is_stand_by = True
    """
    cnt = 0
    for si in c.storageinstance_set.all():
        if (
            si.instance_inner_role == InstanceInnerRole.MASTER
            and si.status == InstanceStatus.RUNNING
            and si.phase == InstancePhase.ONLINE
        ):
            cnt += 1

    if cnt <= 0:
        return [CheckResponse(msg=_("无正常 master"), check_subtype=MetaCheckSubType.ClusterTopo)]


@checker_wrapper
def _cluster_one_standby_slave(
    c: Cluster,
) -> List[CheckResponse]:
    """
    只能有一个 standby slave
    """
    m = []
    for si in c.storageinstance_set.all():
        if si.instance_inner_role == InstanceInnerRole.SLAVE and si.is_stand_by is True:
            m.append(si)

    if len(m) <= 0:
        return [CheckResponse(msg=_("无 standby slave"), check_subtype=MetaCheckSubType.ClusterTopo)]

    if len(m) > 1:
        return [
            CheckResponse(
                msg=_("standby slave 多余 1 个: {}".format(",".join([ele.ip_port for ele in m]))),
                check_subtype=MetaCheckSubType.ClusterTopo,
            )
        ]


@checker_wrapper
def _cluster_standby_slave_status(c: Cluster) -> List[CheckResponse]:
    """
    standby slave 必须正常
    """
    bad = []
    for si in c.storageinstance_set.all():
        if (si.instance_inner_role == InstanceInnerRole.SLAVE and si.is_stand_by is True) and (
            si.status != InstanceStatus.RUNNING or si.phase != InstancePhase.ONLINE
        ):
            bad.append(
                CheckResponse(
                    msg=_("standby slave {} 状态异常: {}, {}".format(si.ip_port, si.status, si.phase)),
                    check_subtype=MetaCheckSubType.ClusterTopo,
                    instance=si,
                )
            )

    return bad

File: dbm-services/mysql/db-tools/dbactuator/pkg/components/mysql/dts_cutover/status.go
/*
 * TencentBlueKing is pleased to support the open source community by making 蓝鲸智云-DB管理系统(BlueKing-BK-DBM) available.
 * Copyright (C) 2017-2023 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at https://opensource.org/licenses/MIT
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

package dts_cutover

import (
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"

	"dbm-services/common/go-pubpkg/logger"
)

// SyncStatus DTS 任务增量同步状态（HTTP status 子集）。
type SyncStatus struct {
	MasterBinlog        string   `json:"master_binlog"`
	MasterBinlogGtid    string   `json:"master_binlog_gtid"`
	SyncerBinlog        string   `json:"syncer_binlog"`
	SyncerBinlogGtid    string   `json:"syncer_binlog_gtid"`
	BlockingDDLs        []string `json:"blocking_ddls"`
	SecondsBehindMaster int      `json:"seconds_behind_master"`
	Synced              bool     `json:"synced"`
}

// TaskStatusItem GET /api/v1/tasks/{task}/status 单条。
type TaskStatusItem struct {
	Name       string      `json:"name"`
	SourceName string      `json:"source_name"`
	Stage      string      `json:"stage"`
	Unit       string      `json:"unit"`
	ErrorMsg   string      `json:"error_msg"`
	SyncStatus *SyncStatus `json:"sync_status"`
}

// TaskStatusListResponse 任务状态列表响应。
type TaskStatusListResponse struct {
	Total int              `json:"total"`
	Data  []TaskStatusItem `json:"data"`
}

// CutoverPositionOutput 供 Flow 落库的位点输出。
type CutoverPositionOutput struct {
	TaskName  string                   `json:"task_name"`
	Sources   []SourcePositionSnapshot `json:"sources"`
	StoppedAt string                   `json:"stopped_at"`
}

// SourcePositionSnapshot 单个 source 的位点快照。
type SourcePositionSnapshot struct {
	SourceName       string `json:"source_name"`
	MasterBinlog     string `json:"master_binlog"`
	MasterBinlogGtid string `json:"master_binlog_gtid"`
	SyncerBinlog     string `json:"syncer_binlog"`
	SyncerBinlogGtid string `json:"syncer_binlog_gtid"`
	SecondsBehind    int    `json:"seconds_behind_master"`
}

// buildMasterAPIURL 拼装 Master OpenAPI URL（本机直连 dts_master_addr）。
func buildMasterAPIURL(dtsMasterAddr, apiPath string) (string, error) {
	addr := strings.TrimSpace(dtsMasterAddr)
	if addr == "" {
		return "", fmt.Errorf("dts_master_addr 为空")
	}
	if !strings.Contains(addr, "://") {
		addr = "http://" + addr
	}
	u, err := url.Parse(addr)
	if err != nil {
		return "", fmt.Errorf("解析 dts_master_addr 失败: %w", err)
	}
	u.Path = apiPath
	u.RawQuery = ""
	return u.String(), nil
}

// FetchTaskStatus HTTP GET http://{dts_master_addr}/api/v1/tasks/{task_name}/status
func FetchTaskStatus(dtsMasterAddr, taskName string, timeoutSec int) (*TaskStatusListResponse, error) {
	if timeoutSec <= 0 {
		timeoutSec = 30
	}
	u, err := buildMasterAPIURL(dtsMasterAddr, fmt.Sprintf("/api/v1/tasks/%s/status", url.PathEscape(taskName)))
	if err != nil {
		return nil, err
	}

	client := &http.Client{Timeout: time.Duration(timeoutSec) * time.Second}
	req, err := http.NewRequest(http.MethodGet, u, nil)
	if err != nil {
		return nil, fmt.Errorf("构造 status 请求失败: %w", err)
	}
	logger.Info("查询 DTS 任务状态: %s", req.URL.String())
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("查询任务状态失败: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("查询任务状态 HTTP %d", resp.StatusCode)
	}
	var out TaskStatusListResponse
	if err = json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return nil, fmt.Errorf("解析任务状态 JSON 失败: %w", err)
	}
	return &out, nil
}

// statusSourceKey 取 status item 的 source 键（优先 source_name）。
func statusSourceKey(item TaskStatusItem) string {
	if item.SourceName != "" {
		return item.SourceName
	}
	return item.Name
}

// BuildLockMasterSnapshots 加锁后从 taskStatus 按 source 记录当时的 master 位点快照。
func BuildLockMasterSnapshots(items []TaskStatusItem) (map[string]BinlogCoord, error) {
	if len(items) == 0 {
		return nil, fmt.Errorf("任务状态为空，无法建立加锁位点快照")
	}
	out := make(map[string]BinlogCoord, len(items))
	for _, item := range items {
		src := statusSourceKey(item)
		if item.SyncStatus == nil {
			return nil, fmt.Errorf("source %s 缺少 sync_status，无法建立加锁位点快照", src)
		}
		coord, ok := ParseBinlogCoord(item.SyncStatus.MasterBinlog)
		if !ok {
			return nil, fmt.Errorf("source %s master_binlog 非法，无法建立加锁位点快照: %q", src, item.SyncStatus.MasterBinlog)
		}
		out[src] = coord
	}
	return out, nil
}

// CheckSnapshotCatchup 持锁复核：所有 source 须 SBM==0、syncer>=加锁 master 快照、无 blocking_ddls。
// 数据一致性由编排侧已完成的 checksum 兜底；实时 master 前进不影响判定。
// 对称约束：本轮 status 必须覆盖 snapshots 中每一个 source，禁止子集追平假成功。
func CheckSnapshotCatchup(items []TaskStatusItem, snapshots map[string]BinlogCoord) error {
	if len(items) == 0 {
		return fmt.Errorf("任务状态为空，无法复核追平")
	}
	if len(snapshots) == 0 {
		return fmt.Errorf("加锁位点快照为空，无法复核追平")
	}
	seen := make(map[string]struct{}, len(items))
	for _, item := range items {
		src := statusSourceKey(item)
		seen[src] = struct{}{}
		if item.ErrorMsg != "" {
			return fmt.Errorf("source %s 任务错误: %s", src, item.ErrorMsg)
		}
		if item.SyncStatus == nil {
			return fmt.Errorf("source %s 缺少 sync_status", src)
		}
		ss := item.SyncStatus
		if len(ss.BlockingDDLs) > 0 {
			return fmt.Errorf("source %s 存在 blocking_ddls: %v", src, ss.BlockingDDLs)
		}
		snap, ok := snapshots[src]
		if !ok {
			return fmt.Errorf("source %s 缺少加锁位点快照", src)
		}
		if !IsCaughtUpToSnapshot(ss.SyncerBinlog, snap, ss.SecondsBehindMaster) {
			return fmt.Errorf(
				"source %s 未追上加锁快照(需 SBM=0 且 syncer>=lock_master): sbm=%d lock_master=(%s, %d) syncer_binlog=%q live_master=%q",
				src, ss.SecondsBehindMaster, snap.File, snap.Position, ss.SyncerBinlog, ss.MasterBinlog,
			)
		}
	}
	for src := range snapshots {
		if _, ok := seen[src]; !ok {
			return fmt.Errorf("加锁快照含 source %s，但本轮 status 未返回该源，拒绝追平假成功", src)
		}
	}
	return nil
}

// BuildPositionOutput 从 status 列表构造输出位点。
func BuildPositionOutput(taskName string, items []TaskStatusItem) CutoverPositionOutput {
	out := CutoverPositionOutput{
		TaskName:  taskName,
		StoppedAt: time.Now().UTC().Format(time.RFC3339),
		Sources:   make([]SourcePositionSnapshot, 0, len(items)),
	}
	for _, item := range items {
		snap := SourcePositionSnapshot{SourceName: item.SourceName}
		if item.SyncStatus != nil {
			snap.MasterBinlog = item.SyncStatus.MasterBinlog
			snap.MasterBinlogGtid = item.SyncStatus.MasterBinlogGtid
			snap.SyncerBinlog = item.SyncStatus.SyncerBinlog
			snap.SyncerBinlogGtid = item.SyncStatus.SyncerBinlogGtid
			snap.SecondsBehind = item.SyncStatus.SecondsBehindMaster
		}
		out.Sources = append(out.Sources, snap)
	}
	return out
}
